### PR TITLE
OpticalFlow: Add support for MinnowBoardMax [v3]

### PR DESCRIPTION
**Pull request by ricardotk**
This PR adds support for OpticalFlow to MinnowBoardMax trying to
leave the OpticalFlow implementation as generic as it already is.

We had to add some format conversion and software crop to the cameras that
do not have this features.

Most cameras do not support NV12 or GREY formats, we are adding in this PR
a conversion from YUYV format, that seems pretty common in cameras, to GREY
format (since we do not use Cb and Cr data on OpticalFlow).

Also we are adding a software crop for 8bpp images such as GREY format. For the
same reason, most cameras do not have support for overlaying (crop, resize and so on).

This PR is still a WIP version since it still need to be tested on a flying drone, but the tests I had made with a camera seemed to be correct.

* Differences from v1:
  * Changed `crop_image_8bpp()` algorithm
  * Correct check to see if crop is supported
  * Remove unused `get_fd()` function
  * Adopt 80~100 columns coding style

* Differences from v2:
  * Add `get_pixel_format()` function to query the supported formats from the camera we are using
  * `crop_8bpp()` and `yuyv_to_grey()` function names changed
  * Changed the usage of `alloca()` to `malloc()` to allocate the buffers

### CODE REVIEW

File: libraries/AP_HAL/AP_HAL_Boards.h [hunk 1]
```diff
@@ -250,6 +250,22 @@
...
+#define HAL_OPTFLOW_ONBOARD_CROP_HEIGHT 240
+#define HAL_OPTFLOW_ONBOARD_NBUFS 8;
+#define HAL_FLOW_PX4_MAX_FLOW_PIXEL 4
+#define HAL_FLOW_PX4_BOTTOM_FLOW_FEATURE_THRESHOLD 30
+#define HAL_FLOW_PX4_BOTTOM_FLOW_VALUE_THRESHOLD 5000
+#define HAL_PARAM_DEFAULTS_PATH "/etc/arducopter/minlure.parm"
+/* ELP-USBFHD01M-L21
+ * focal length 2.1 mm, pixel size 3 um
+ * 240x240 crop rescaled to 64x64 */
+#define HAL_FLOW_PX4_FOCAL_LENGTH_MILLIPX (3.0 / (2.1 * 2.0 * 240 / 64))
```
Comment (jberaud) [inline libraries/AP_HAL/AP_HAL_Boards.h]: Finally you had a 2x2 binning?

**Comment by ricardotk** [comment]
@lucasdemarchi Would you mind taking a look?

**Comment by jberaud** [comment]
@ricardotk you haven't implemented the use of v4l2 enum_fmt if I am not mistaken

**Comment by ricardotk** [comment]
>you haven't implemented the use of v4l2 enum_fmt if I am not mistaken

`get_pixel_formats()` does the job

**Comment by jberaud** [comment]
My mistake. Sorry.

**Comment by jberaud** [comment]
Ok for me

File: libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp [hunk 2]
```diff
@@ -90,18 +77,68 @@ void OpticalFlow_Onboard::init(AP_HAL::OpticalFlow::Gyro_Cb get_gyro)
...
+
+        /* if V4L2_PIX_FMT_YUYV format is found we still iterate through
+         * the vector since the other formats need no conversions. */
+        if (px_fmt == V4L2_PIX_FMT_YUYV) {
+            _format = px_fmt;
+        }
+    }
+
+    if (!_format) {
+        AP_HAL::panic("OpticalFlow_Onboard: no supported format available");
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp]: there's already a panic below if the format doesn't match one of the supported formats

File: libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp [hunk 3]
```diff
@@ -247,5 +346,13 @@ void OpticalFlow_Onboard::_run_optflow()
...
         _last_video_frame = video_frame;
     }
+
+    if (convert_buffer) {
+        free(convert_buffer);
+    }
+
+    if (crop_buffer) {
+        free(crop_buffer);
+    }
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/OpticalFlow_Onboard.cpp]: as far as I can see, if we don't have a way to top the thread this code is actually never executed.

**Comment by lucasdemarchi** [comment]
Applied. Any new improvement can be done on top.

**Pull request closed without merging**